src/kbmod/search/raw_image_eigen.cpp

```diff
@@ -0,0 +1,767 @@
+  // Image zeroed = (array.array() == NO_DATA).select(0, array.array());
+  // result(i, j) = zeroed.block<krows, kcols>(i, j).cwiseProduct(kernel).sum()
+  void RawImageEigen::convolve_cpu(PSF& psf) {
+    Image result = Image::Zero(height, width);
+
+    // cast the PSF kernel as an Eigen Matrix for cleanliness
+    // remember to undo later. Also, what is the PSF dimension?
+    Eigen::Map<Eigen::MatrixXf> kernel_raw(psf.data(), 1, psf.get_size());
+    auto kernel = kernel_raw.reshaped(psf.get_dim(), psf.get_dim());
+
+    const int psf_rad = psf.get_radius();
+    const float psf_total = psf.get_sum();
+
+    for (int y = 0; y < height; ++y) {
+      for (int x = 0; x < width; ++x) {
+        // Pixels with NO_DATA remain NO_DATA.
+        if (image(x, y) == NO_DATA){
+          result(x, y) = NO_DATA;
+          continue;
+        }
+
+        float sum = 0.0;
+        float psf_portion = 0.0;
+        for (int j = -psf_rad; j <= psf_rad; j++) {
+          for (int i = -psf_rad; i <= psf_rad; i++) {
+            Index idx(x+i, y+j);
+            if (idx.is_within(width, height) &&  image(idx.i, idx.j) != NO_DATA){
+              psf_portion += kernel(i+psf_rad, j+psf_rad);
+              sum += image(idx.i, idx.j) * kernel(i+psf_rad, j+psf_rad);
+            }
+          } // for i
+        } // for j
+        result(x, y) = (sum * psf_total) / psf_portion;
+      } // for x
+    } // for y
+
+    // avoid copying the results back
+    image = std::move(result);
+  }
+
+
+#ifdef HAVE_CUDA
+  // Performs convolution between an image represented as an array of floats
+  // and a PSF on a GPU device.
+  extern "C" void deviceConvolve(float* source_img, float* result_img,
+                                 int width, int height, float* psf_kernel,
+                                 int psf_size, int psf_dim, int psf_radius,
+                                 float psf_sum);
+#endif
+
+
+  void RawImageEigen::convolve(PSF psf) {
+#ifdef HAVE_CUDA
+    deviceConvolve(image.data(), image.data(),
+                   get_width(), get_height(), psf.data(),
+                   psf.get_size(), psf.get_dim(), psf.get_radius(), psf.get_sum());
+#else
+    convolve_cpu(psf);
+#endif
+  }
+
+
+  void RawImageEigen::apply_bitmask(const ImageI& bitmask) {
+    image = (bitmask.array() == 1).select(NO_DATA, image);
+  }
+
+
+  void RawImageEigen::apply_mask1(const ImageI& mask, const int flags) {
+    // I'm not sure a copy is avoidable here, we can pass a non-const Ref
+    // and then edit the mask in-place, but that would edit the Python-side
+    // numpy array as well....
+    auto bitmask = mask.unaryExpr( [&flags](int x){return flags & x;} );
+    apply_bitmask(bitmask);
+  }
+
+
+  // Imma be honest, this function makes no sense to me... I think I transcribed
+  // it right, but I'm not sure? If a flag is in mask, it's masked unless it's in
+  // exceptions? But why do we have two lists? Is the example above better than this?
```

Comment:
As we discussed earlier (but documenting here for others), I this exceptions operates at the level of the entire string where flags are per-bit, so you could do:

Mask if any of the bits {1, 2, 4, 5} are set, EXCEPT if it is the pattern "11010" or "11001"